programs/steward/src/instructions/epoch_maintenance.rs

```diff
@@ -0,0 +1,95 @@
+use crate::{
+    errors::StewardError,
+    utils::{
+        check_validator_list_has_stake_status, get_stake_pool, get_validator_list_length, StakePool,
+    },
+    Config, StewardStateAccount,
+};
+use anchor_lang::prelude::*;
+use spl_stake_pool::state::StakeStatus;
+
+#[derive(Accounts)]
+pub struct EpochMaintenance<'info> {
+    pub config: AccountLoader<'info, Config>,
+
+    #[account(
+        mut,
+        seeds = [StewardStateAccount::SEED, config.key().as_ref()],
+        bump
+    )]
+    pub state_account: AccountLoader<'info, StewardStateAccount>,
+
+    #[account(mut, address = stake_pool.validator_list)]
+    pub validator_list: AccountInfo<'info>,
+
+    #[account(
+        address = get_stake_pool(&config)?
+    )]
+    pub stake_pool: Account<'info, StakePool>,
+}
+
+/// Runs maintenance tasks at the start of each epoch, needs to be run multiple times
+/// Routines:
+/// - Remove delinquent validators
+pub fn handler(
+    ctx: Context<EpochMaintenance>,
+    validator_index_to_remove: Option<usize>,
+) -> Result<()> {
+    let stake_pool = &ctx.accounts.stake_pool;
+    let mut state_account = ctx.accounts.state_account.load_mut()?;
+
+    let clock = Clock::get()?;
+
+    require!(
+        clock.epoch == stake_pool.last_update_epoch,
+        StewardError::StakePoolNotUpdated
+    );
+
+    if (!state_account.state.checked_validators_removed_from_list).into() {
+        // Ensure there are no validators in the list that have not been removed, that should be
+        require!(
+            !check_validator_list_has_stake_status(
+                &ctx.accounts.validator_list,
+                StakeStatus::ReadyForRemoval
+            )?,
+            StewardError::ValidatorsHaveNotBeenRemoved
+        );
```

Comment:
would be curious to see the CU cost on a MAX_VALIDATORS validator list, if it's not too big, might be worth just running this function for every IX to be able to get rid of the `checked_validators_removed` field

Reply:
Taking out for now - of need be we can add back in later